### PR TITLE
Perform data reset for users already affected by the encryption issues (EXPOSUREAPP-1851/EXPOSUREAPP-2958)

### PR DESCRIPTION
**Pull request by d4rken**
~~**DO NOT MERGE** but review :wink:~~
~~This still needs the final translations for the info dialog after a reset.~~

## Description
The mitigation from #1235 only helps users not already affected by the issue (see #642).

Existing users will still keep crashing. There is graceful recovery because the encryption key is gone.
They have to reset their app data. We can do this for them. If the user has automatic updates enabled in Google Play, there may be users who have already been affected, don't know it yet, and where it will be fixed automatically too. They will only see a dialog when they open their app the next time, informing them about what happened (see mock screenshot).

The result should be that the app automatically starts working again if an affected user updates it. We will loose the encrypted data, but the automatic tasks should be able to work again and thus the app  perform it's duty.

**When do we do it?**
* It is a `GeneralSecurityException("decryption failed")`
* The encrypted shared preferences actually exist (otherwise it's a different error cause)
* This is the first error we encountered when creating the encrypted shared preferences, which means that there was no other error before we encountered this one. This allows us to only perform the reset for users upgrading from <1.4.0 to 1.4.0+, as we don't want to reset for any other case potentially hiding other issues.

**What do we do?**
* If the specific error case is matched
* Deleted the encrypted shared preferences
* Delete the encrypted database
* Retry creation of the encrypted shared preferences

## How to test
* Open the app
* Review the code
* Review the tests, are all edge cases covered?
* Install the app, complete onboarding, insert a `throw GeneralSecurityException("decryption failed")` ~~into the `EncryptedPreferencesFactory`~~ actually, insert it into the [SecurityHelper](https://github.com/corona-warn-app/cwa-app-android/pull/1256/files#diff-d05886a3ee6f8f08529a36f8e9fb936bR49), otherwise we keep throwing. Install the new version without data reset. You should see the app reset and then display the dialog, once.

### CODE REVIEW

File: Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/ui/main/MainFragment.kt [hunk 1]
```diff
@@ -22,6 +22,8 @@ import de.rki.coronawarnapp.ui.viewmodel.SubmissionViewModel
 import de.rki.coronawarnapp.ui.viewmodel.TracingViewModel
 import de.rki.coronawarnapp.util.DialogHelper
 import de.rki.coronawarnapp.util.ExternalActionHelper
+import de.rki.coronawarnapp.util.di.AppInjector
```
Comment (harambasicluka) [inline Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/ui/main/MainFragment.kt:25]: Why is this the Fragment in `.../deviceForTesters/...`?

Comment (d4rken) [reply Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/ui/main/MainFragment.kt:25]: :rage: that's the split code issue again, good spot. Need to add it for both fragments `device` and `deviceForTesters`.

Comment (harambasicluka) [reply Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/ui/main/MainFragment.kt:25]: It's about the file not about the line.

File: Corona-Warn-App/src/device/java/de.rki.coronawarnapp/ui/main/MainFragment.kt [hunk 2]
```diff
@@ -68,6 +74,19 @@ class MainFragment : Fragment() {
         showOneTimeTracingExplanationDialog()
     }
 
+    override fun onActivityCreated(savedInstanceState: Bundle?) {
+        super.onActivityCreated(savedInstanceState)
+
+        if (errorResetTool.isResetNoticeToBeShown) {
```
Comment (ralfgehrer) [inline Corona-Warn-App/src/device/java/de.rki.coronawarnapp/ui/main/MainFragment.kt:80]: How about moving this block to a dedicated function "showResetNoticeIfNeeded"

File: Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/ui/main/MainFragment.kt [hunk 3]
```diff
@@ -67,6 +73,19 @@ class MainFragment : Fragment() {
         showOneTimeTracingExplanationDialog()
     }
 
+    override fun onActivityCreated(savedInstanceState: Bundle?) {
+        super.onActivityCreated(savedInstanceState)
+
+        if (errorResetTool.isResetNoticeToBeShown) {
```
Comment (ralfgehrer) [inline Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/ui/main/MainFragment.kt:79]: How about moving this block to a dedicated function "showResetNoticeIfNeeded"

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/security/EncryptionErrorResetTool.kt [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+        } catch (e: Exception) {
+            // If anything goes wrong, we return false, so that our caller can rethrow the original error.
+            Timber.e(e, "Error while performing the reset.")
+            false
+        }
+    }
+
+    private fun performReset(): Boolean {
+        // Delete encrypted shared preferences file
+
```
Comment (ralfgehrer) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/security/EncryptionErrorResetTool.kt:110]: no empty line plz

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/security/EncryptionErrorResetTool.kt [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+            return false
+        }
+
+        resetPerformedAt = Instant.now()
+        isResetNoticeToBeShown = true
+
+        return true
+    }
+
+    companion object {
```
Comment (ralfgehrer) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/security/EncryptionErrorResetTool.kt:132]: to top of class

**Review by ralfgehrer — DISMISSED** [review]

File: Corona-Warn-App/src/device/java/de.rki.coronawarnapp/ui/main/MainFragment.kt (hunk 2 above)
Comment (chris-cwa) [reply Corona-Warn-App/src/device/java/de.rki.coronawarnapp/ui/main/MainFragment.kt:80]: <personal opinion>
I don't like methods that are only called from one location. Use doc to make more readable.
</personal opinion>

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/security/EncryptionErrorResetTool.kt [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+    private var isResetWindowConsumed: Boolean
+        get() = prefs.getBoolean(PKEY_EA1851_WAS_WINDOW_CONSUMED, false)
+        set(value) = prefs.edit {
+            putBoolean(PKEY_EA1851_WAS_WINDOW_CONSUMED, value)
+        }
+
+    private var resetPerformedAt: Instant?
+        get() {
+            val performedAt = prefs.getLong(PKEY_EA1851_RESET_PERFORMED_AT, -1)
+            return if (performedAt == -1L) null else Instant.ofEpochMilli(performedAt)
```
Comment (chris-cwa) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/security/EncryptionErrorResetTool.kt:49]: use Kotlin style
```suggestion
            prefs.getLong(PKEY_EA1851_RESET_PERFORMED_AT, -1).also{
                return if (it == -1L) null else Instant.ofEpochMilli(performedAt)
            }
```

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/security/SecurityHelper.kt [hunk 7]
```diff
@@ -23,26 +23,44 @@ import android.annotation.SuppressLint
...
 import de.rki.coronawarnapp.util.security.SecurityConstants.DB_PASSWORD_MAX_LENGTH
 import de.rki.coronawarnapp.util.security.SecurityConstants.DB_PASSWORD_MIN_LENGTH
 import de.rki.coronawarnapp.util.security.SecurityConstants.ENCRYPTED_SHARED_PREFERENCES_FILE
+import timber.log.Timber
 import java.security.SecureRandom
 
 /**
  * Key Store and Password Access
  */
 object SecurityHelper {
```
Comment (chris-cwa) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/security/SecurityHelper.kt:40]: this should be an injectable class, ideally with interface (maybe out of scope for this PR -> 1 more for refactoring epic)

**Review by chris-cwa — DISMISSED** [review]

File: Corona-Warn-App/src/device/java/de.rki.coronawarnapp/ui/main/MainFragment.kt (hunk 2 above)
Comment (chris-cwa) [reply Corona-Warn-App/src/device/java/de.rki.coronawarnapp/ui/main/MainFragment.kt:80]: > How about moving this block to a dedicated function "showResetNoticeIfNeeded"

I don't like methods that are only called from one location. Use doc to make more readable.

File: Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/ui/main/MainFragment.kt (hunk 3 above)
Comment (chris-cwa) [reply Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/ui/main/MainFragment.kt:79]: > How about moving this block to a dedicated function "showResetNoticeIfNeeded"

I don't like methods that are only called from one location. Use doc to make more readable.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/security/EncryptionErrorResetTool.kt (hunk 6 above)
Comment (chris-cwa) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/security/EncryptionErrorResetTool.kt:49]: define const INVALID = -1L to make it more clear?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/security/EncryptionErrorResetTool.kt (hunk 5 above)
Comment (d4rken) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/security/EncryptionErrorResetTool.kt:132]: We attempt to follow the official Kotlin conventions, which say `companion object` last: https://kotlinlang.org/docs/reference/coding-conventions.html

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/security/EncryptionErrorResetTool.kt (hunk 6 above)
Comment (d4rken) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/security/EncryptionErrorResetTool.kt:49]: Agree, that's nicer, will do! Btw we need to use `let` not `also`. Only `let` modifies the returned result.

File: Corona-Warn-App/src/device/java/de.rki.coronawarnapp/ui/main/MainFragment.kt (hunk 2 above)
Comment (d4rken) [reply Corona-Warn-App/src/device/java/de.rki.coronawarnapp/ui/main/MainFragment.kt:80]: I concur with @chris-cwa here. I also think `errorResetTool.isResetNoticeToBeShown` is descriptive enough that no comment is warranted. But let's revisit the MainFragment for refactoring in one of the next sprints in any case!

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/security/SecurityHelper.kt (hunk 7 above)
Comment (d4rken) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/security/SecurityHelper.kt:40]: Yep out of scope here, would require too many changes, but should be refactored as part of our no-more-singletons tickets.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/security/EncryptionErrorResetTool.kt (hunk 6 above)
Comment (d4rken) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/security/EncryptionErrorResetTool.kt:49]: Not here, it's so close to each other. I would have also used `null` for the check but `getLong` doesn't support `Long?` as default value.

File: Corona-Warn-App/config/detekt.yml [hunk 8]
```diff
@@ -568,7 +568,7 @@ style:
     active: false
   ReturnCount:
     active: true
-    max: 2
+    max: 6
```
[the remark this answers is not in the document]
Comment (d4rken) [reply Corona-Warn-App/config/detekt.yml:571]: I agree but I also think we need to raise this as it is  improving code readability.

Changing the `tryResetIfNecessary` method to use less returns would reduce readability imho.

While we could remove this change, and then add `Supress(ReturnCount)` to the `tryResetIfNecessary`, then we would have to remove that again when we change the `ReturnCount` value in an extra PR.

So unless someone is opposed to increasing the `ReturnCount` value itself, I would keep it as is.

**Review by AlexanderAlferov — APPROVED** [review]

**Review by ralfgehrer — APPROVED** [review]

**Review by chris-cwa — APPROVED** [review]
